src/value.mts

```diff
@@ -22,149 +21,268 @@ import { EnvironmentRecord } from './environment.mjs';
 import { Completion, X } from './completion.mjs';
 import { ValueMap, OutOfRange, callable } from './helpers.mjs';
 import type { PrivateElementRecord } from './runtime-semantics/MethodDefinitionEvaluation.mjs';
+import type { ParseNode } from './parser/ParseNode.mjs';
 
+let createStringValue: (value: string) => StringValue;
+let createNumberValue: (value: number) => NumberValue;
+let createBigIntValue: (value: bigint) => BigIntValue;
+
+abstract class BaseValue {
+  static declare readonly null: NullValue;
+  static declare readonly undefined: UndefinedValue;
+  static declare readonly true: BooleanValue<true>;
+  static declare readonly false: BooleanValue<false>;
+  abstract type: Value['type'];
+}
 
-// @ts-expect-error callable class
-export declare function Value(value: string): StringValue; // @ts-expect-error callable class
-export declare function Value(value: number): NumberValue; // @ts-expect-error callable class
-export declare function Value(value: bigint): BigIntValue; // @ts-expect-error callable class
-export declare function Value(value: undefined): UndefinedValue; // @ts-expect-error callable class
-export declare function Value(value: null): NullValue; // @ts-expect-error callable class
-// TODO(ts): define a FunctionObjectValue type.
-export declare function Value(value: (...args: unknown[]) => unknown): ObjectValue; // @ts-expect-error callable class
-export declare function Value(value: string | number): StringValue | NumberValue;
 /** https://tc39.es/ecma262/#sec-ecmascript-language-types */
-export @callable((_target, _thisArg, [value]) => {
-  if (value === null) {
-    return Value.null;
-  }
-  switch (typeof value) {
-    case 'undefined':
+export type Value =
+  | UndefinedValue
+  | NullValue
+  | BooleanValue
+  | StringValue
+  | SymbolValue
+  | NumberValue
+  | BigIntValue
+  | PrivateName
```

Comment:
Is PrivateName a Value?

Reply:
It is already defined as a `Value` in engine262, this doesn't change that.

Reply:
You are correct that it doesn't make sense for it to be considered a `Value`, if `Value` is to mean the same thing as "ECMAScript Language Value" in the spec, so I've removed it from `Value` in the most recent change.